<a href="https://colab.research.google.com/github/OswaldVc/Datos/blob/main/RPN1/Regrelogistica.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Regresion Logistica

la regresión logística es un tipo de análisis de regresión utilizado para predecir el resultado de una variable categórica (una variable que puede adoptar un número limitado de categorías, sobre todo binomiales) en función de las variables independientes o predictoras.  Es decir predecir si una futura compra la hara un hombre o una mujer dependiendo de otras variables como lo son edad, ingresos, horas libres, monto de la ultima compra.


Como este responde a un modelo de predicción la base de datos suele dividirse en dos grupos, un primer grupo para entramiento del modelo y un segundo grupo para validacion o preubas del modelo.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import statsmodels.api as sm
import pandas as pd

In [ ]:
df=pd.read_csv('logit_train1.csv',index_col=0)

FileNotFoundError: ignored

In [ ]:
df.head()

gmat  gpa  work_experience  admitted
16   580  2.7                4         0
33   660  3.3                6         1
8    740  3.3                5         1
38   590  1.7                4         0
32   660  4.0                4         1

Definimimos el conjunto de entrenamiento, esta corresponde a puntajes obtenidos por un conjuto de aspirantes a una beca. La variable a predecir (y) es la variable "admitted" y las demas variables seran las variables predictoras.
Por lo tanto definimos cada uno de estos conjuntos

In [ ]:
xtrain =df[['gmat', 'gpa', 'work_experience']]
ytrain=df[['admitted']]

Entrenamiento del modelo

In [ ]:
log_reg=sm.Logit(ytrain, xtrain).fit()

Optimization terminated successfully.
         Current function value: 0.352707
         Iterations 8


In [ ]:
#Revisaremos todos los indicadores y parametros del modelo 
print(log_reg.summary())


                           Logit Regression Results                           
Dep. Variable:               admitted   No. Observations:                   30
Model:                          Logit   Df Residuals:                       27
Method:                           MLE   Df Model:                            2
Date:                Mon, 19 Apr 2021   Pseudo R-squ.:                  0.4912
Time:                        17:27:07   Log-Likelihood:                -10.581
converged:                       True   LL-Null:                       -20.794
Covariance Type:            nonrobust   LLR p-value:                 3.668e-05
                      coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------
gmat               -0.0262      0.011     -2.383      0.017      -0.048      -0.005
gpa                 3.9422      1.964      2.007      0.045       0.092       7.792
work_experience     1.1983      

Ahora bien el modelo resultante es  admitted=-0.0262*gmat+3.9422 *gpa+1.1983*work_experience. ESte NO es un buen modelo porque si observamos  Pseudo R-squ. es de 0.49 y este debe ser cercano a 1.

Vamos a validar el modelo para eso usaremos el cnjunto de datos de prueba y revisar las predicciones que se obtienen.

In [ ]:
# loading the testing dataset
df = pd.read_csv('logit_test1.csv', index_col = 0)

# defining the dependent and independent variables
Xtest = df[['gmat', 'gpa', 'work_experience']]
ytest = df['admitted']

# performing predictions on the test datdaset
yhat = log_reg.predict(Xtest)
prediction = list(map(round, yhat))

# comparing original and predicted values of y
print('Acutal values', list(ytest.values))
print('Predictions :', prediction)


Acutal values [0, 0, 0, 0, 0, 1, 1, 0, 1, 1]
Predictions : [0, 0, 0, 0, 0, 0, 0, 0, 1, 1]


Los vectore resultantes nos muestran los valores de los datos y valores predecidos por el modelo como observamos falla en algunos, puesto que como ya se dijo anteriormente no es un buen modelo.  Es importante tener en cuenta que un modelo perfecto no extiste y siempre existiran datos que no  pueda predecir con total exactitud.

Vamos a revisar otras metricas sobre las predicciones, para esto usaremos el accuracy ( Leer https://www.iartificial.net/precision-recall-f1-accuracy-en-clasificacion/#:~:text=Accuracy%20(Exactitud),-La%20exactitud%20(accuracy&text=Es%20decir%2C%20el%20modelo%20acierta%20el%2075%25%20de%20las%20veces.&text=De%20hecho%2C%20si%20tuvi%C3%A9semos%20un,su%20accuracy%20ser%C3%ADa%20del%2080%25.)   El accuracy es la exactitud del modelo es decir muestra un porcentaje de veces que lo hace exacto, por lo tanto se busca que este sea lo mas alto posible  o cercano a 1.

Tambien existen otras metricas como las matriz de confusion.

In [ ]:
from sklearn.metrics import (confusion_matrix, accuracy_score)

 # accuracy score of the model
print('Test accuracy = ', accuracy_score(ytest, prediction))

# confusion matrix
cm = confusion_matrix(ytest, prediction )
print ("Confusion Matrix : \n", cm) 

Test accuracy =  0.8
Confusion Matrix : 
 [[6 0]
 [2 2]]
